### PR TITLE
plugin/forward: add dnstap support

### PR DESCRIPTION
**Pull request by 2buisson**
### 2. Which issues (if any) are related?

* https://github.com/coredns/coredns/issues/1448

### CODE REVIEW

File: plugin/forward/connect.go [hunk 1]
```diff
@@ -114,12 +114,18 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts options
 
 	conn.SetReadDeadline(time.Now().Add(p.readTimeout()))
 	ret, err := conn.ReadMsg()
+
+	taperr := logToDnstap(ctx, p.transport.addr, proto, state.Req, ret, start)
```
Comment (rdrozhdzh) [inline plugin/forward/connect.go]: `proto` here does not always corresponds to actual protocol, e.g. `transport.Dial()` can override to use `tcp-tls` if TLS was configured. Perhaps, we could address this issue in a separate PR

Comment (rdrozhdzh) [inline plugin/forward/connect.go]: you can pass `reqTime` instead of `start` as last argument

File: plugin/forward/dnstap.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/miekg/dns"
+)
+
+func logToDnstap(ctx context.Context, host, proto string, req, ret *dns.Msg, start time.Time) error {
+	t := dnstap.TapperFromContext(ctx)
+	if t == nil {
+		return nil
+	}
+
+	b := msg.New().Time(start).HostPort(host)
```
Comment (rdrozhdzh) [inline plugin/forward/dnstap.go]: Consider renaming `Time()` -> `WithTime()`, `HostPort()` -> `WithHostPort()`. Current names look like getters.

File: plugin/forward/dnstap_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		ret.SetReply(r)
+		ret.Answer = append(ret.Answer, test.A("example.org. IN A 127.0.0.1"))
+		w.WriteMsg(ret)
+	})
+	defer s.Close()
+
+	c := caddy.NewTestController("dns", "forward . "+s.Addr)
+	f, err := parseForward(c)
+	if err != nil {
+		t.Errorf("Failed to create forwarder: %s", err)
```
Comment (rdrozhdzh) [inline plugin/forward/dnstap_test.go:42]: consider using `t.Fatalf()` here, no sense to continue if forwarder has not been created

File: plugin/forward/dnstap_test.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	defer f.OnShutdown()
+
+	dtapHandler.Next = f
+
+	m := new(dns.Msg)
+	m.SetQuestion("example.org.", dns.TypeA)
+	rec := dnstest.NewRecorder(&test.ResponseWriter{})
+
+	if _, err := dtapHandler.ServeDNS(context.TODO(), rec, m); err != nil {
+		t.Fatal("Expected to receive reply, but didn't")
```
Comment (rdrozhdzh) [inline plugin/forward/dnstap_test.go:54]: consider using `t.Errorf()` here and below where we can safely do more checks

File: plugin/forward/dnstap.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		b.SocketProto = tap.SocketProtocol_TCP
+	} else {
+		b.SocketProto = tap.SocketProtocol_UDP
+	}
+
+	if t.Pack() {
+		b.Msg(req)
+	}
+
+	m, err := b.ToOutsideQuery(tap.Message_FORWARDER_QUERY)
```
Comment (rdrozhdzh) [inline plugin/forward/dnstap.go:27]: consider adding a config option to dnstap plugin which defines which types of dnstap messages should be logged and which ones should not, e.g.
```
dnstap <socket> {
    messages CQ CR FQ FR AR RQ
}
```
Other plugins could send (or not to send) particular dnstap messages based on this configuration

Comment (2buisson) [reply plugin/forward/dnstap.go:27]: I need to think about this

Comment (rdrozhdzh) [reply plugin/forward/dnstap.go:27]: This could be done in a separate PR

File: plugin/forward/dnstap_test.go (hunk 3 above)
Comment (2buisson) [reply plugin/forward/dnstap_test.go:42]: Copy-paste from https://github.com/coredns/coredns/blob/master/plugin/forward/proxy_test.go#L53

File: plugin/forward/dnstap_test.go (hunk 4 above)
Comment (2buisson) [reply plugin/forward/dnstap_test.go:54]: Also copy-paste

**Comment by miekg** [comment]
fixed via 403d42ce

**Pull request closed without merging**